### PR TITLE
Enrolling a new student deletes previous student in the list #6534

### LINKED ISSUE #6534: Enrolling a new student deletes previous student in the list
Environment : live server at V5.94

A user has reported that adding a new student to the course deletes the student immediately above the new student (in alphabetical order). I tried using the same course and the behavior is confirmed for the student user wanted to add.
However, I was able to add an entirely new student to the same course without a problem. So I wonder if the problem is related to the fact that the student was already in the course before and was deleted but some of his responses were left over (should have been deleted, but wasn't).

### PR DESCRIPTION
**Pull request by whipermr5**
Fixes #6534

**Outline of Solution**

> Tell us how you solved the issue

- Upon email address update, delete the `CourseStudent` entity and create a new one (this will generate a new `id` that corresponds to the new email).
- Reset Google IDs and resend emails.

### CODE REVIEW

**Comment by whipermr5** [comment]
Ready for review!

**Comment by damithc** [comment]
It's good if there is a test case that fails before this fix is applied.

**Comment by whipermr5** [comment]
@damithc Done!

File: src/main/java/teammates/common/datatransfer/attributes/StudentAttributes.java [hunk 1]
```diff
@@ -372,4 +372,14 @@ public boolean isSectionChanged(StudentAttributes originalStudentAttribute) {
     public boolean isTeamChanged(StudentAttributes originalStudentAttribute) {
         return this.team != null && !this.team.equals(originalStudentAttribute.team);
     }
+    
+    /**
+     * Checks whether the edit form of student has changed the email value.
```
Comment (damithc) [inline src/main/java/teammates/common/datatransfer/attributes/StudentAttributes.java]: This lower level class should not be aware of an 'edit form'.

File: src/main/java/teammates/common/datatransfer/attributes/StudentAttributes.java [hunk 2]
```diff
@@ -372,4 +372,14 @@ public boolean isSectionChanged(StudentAttributes originalStudentAttribute) {
...
+     * 
+     * @param originalStudentAttribute
```
Comment (damithc) [inline src/main/java/teammates/common/datatransfer/attributes/StudentAttributes.java]: If there is nothing else to say about the parameter, remove the line altogether.

File: src/main/java/teammates/common/datatransfer/attributes/StudentAttributes.java [hunk 3]
```diff
@@ -372,4 +372,14 @@ public boolean isSectionChanged(StudentAttributes originalStudentAttribute) {
...
+     * @return true if email value has changed from its original value.
```
Comment (damithc) [inline src/main/java/teammates/common/datatransfer/attributes/StudentAttributes.java]: This line can be removed and the overview sentence can be changed to `Returns true if email value has changed from its original value`

File: src/main/java/teammates/ui/controller/InstructorCourseStudentDetailsEditSaveAction.java [hunk 4]
```diff
@@ -57,13 +57,21 @@ public ActionResult execute() throws EntityDoesNotExistException {
...
                 logic.validateSectionsAndTeams(Arrays.asList(student), courseId);
             } else if (isTeamChanged) {
                 logic.validateTeams(Arrays.asList(student), courseId);
             }
             
             logic.updateStudent(studentEmail, student);
+            
+            if (isEmailChanged) {
+                logic.resetStudentGoogleId(student.email, courseId);
+                taskQueuer.scheduleCourseRegistrationInviteToStudent(courseId, student.email, true);
```
Comment (damithc) [inline src/main/java/teammates/ui/controller/InstructorCourseStudentDetailsEditSaveAction.java]: Wouldn't this send an email to student even before they are supposed to receive such emails i.e. even before any session is open

File: src/main/java/teammates/storage/api/StudentsDb.java [hunk 5]
```diff
@@ -403,31 +404,43 @@ public void updateStudent(String courseId, String email, String newName,
...
-                putDocument(new StudentAttributes(courseStudent));
+                // Set true to prevent changes to last update timestamp
+                courseStudent.keepUpdateTimestamp = keepUpdateTimestamp;
             }
-        
-            // Set true to prevent changes to last update timestamp
-            courseStudent.keepUpdateTimestamp = keepUpdateTimestamp;
-            
-            log.info(Const.SystemParams.COURSE_BACKUP_LOG_MSG + courseId);
         }
```
Comment (damithc) [inline src/main/java/teammates/storage/api/StudentsDb.java]: This method looks a bit too long. May be refactor in a separate PR?

File: src/main/java/teammates/common/datatransfer/attributes/StudentAttributes.java (hunk 1 above)
Comment (whipermr5) [inline src/main/java/teammates/common/datatransfer/attributes/StudentAttributes.java]: Okay. Can I edit the doc comments for the two methods above (`isSectionChanged ` and `isTeamChanged`) in the same PR? They suffer from the same phrasing. Or must I open a new first-timer issue (how strict is our policy)?

Comment (damithc) [inline src/main/java/teammates/common/datatransfer/attributes/StudentAttributes.java]: I think better to do it as a firsttimer issue (we are short of those anyway :-p)

File: src/main/java/teammates/ui/controller/InstructorCourseStudentDetailsEditSaveAction.java (hunk 4 above)
Comment (whipermr5) [inline src/main/java/teammates/ui/controller/InstructorCourseStudentDetailsEditSaveAction.java]: Oops; thank you for pointing this out! Would sending the email only if the student has already joined the course be a good solution?

File: src/main/java/teammates/storage/api/StudentsDb.java (hunk 5 above)
Comment (whipermr5) [inline src/main/java/teammates/storage/api/StudentsDb.java]: Got it!

File: src/main/java/teammates/ui/controller/InstructorCourseStudentDetailsEditSaveAction.java (hunk 4 above)
Comment (damithc) [inline src/main/java/teammates/ui/controller/InstructorCourseStudentDetailsEditSaveAction.java]: I remember seeing an ongoing PR for sending emails when the email is changed. Currently, we don't send out any emails when the email is changed. Students rarely join courses but they need the updated link to submit responses or view results. Perhaps that can be done as a separate PR?

Comment (whipermr5) [inline src/main/java/teammates/ui/controller/InstructorCourseStudentDetailsEditSaveAction.java]: Okay will remove the line first then!

**Comment by whipermr5** [comment]
Ready for review

File: src/test/java/teammates/test/cases/browsertests/InstructorCourseStudentDetailsEditPageUiTest.java [hunk 6]
```diff
@@ -134,9 +134,15 @@ public void testEditAction() {
...
         assertEquals("New team", student.team);
-        assertEquals(testData.students.get("registeredStudent").googleId, student.googleId);
         assertEquals("newemail@gmail.tmt", student.email);
         assertEquals("New comments", student.comments);
+        assertTrue(student.googleId.isEmpty()); // Due to Google ID reset
+        
+        // Verify adding the original student again does not overwrite the edited entity
+        BackDoor.createStudent(testData.students.get("registeredStudent"));
+        student = BackDoor.getStudent(testData.courses.get("CCSDEditUiT.CS2104").getId(),
+                                      "newemail@gmail.tmt");
```
Comment (damithc) [inline src/test/java/teammates/test/cases/browsertests/InstructorCourseStudentDetailsEditPageUiTest.java]: Not related to this PR: `newemail@gmail.tmt` is used a few times in this test. Should be declared as a local variable. Furthermore, the whole `// Verify data` block should be abstracted as a util method.

File: src/main/java/teammates/storage/api/StudentsDb.java [hunk 7]
```diff
@@ -403,31 +404,43 @@ public void updateStudent(String courseId, String email, String newName,
...
-    
-            courseStudent.setEmail(newEmail);
-            courseStudent.setName(newName);
-            courseStudent.setLastName(StringHelper.splitName(newName)[1]);
-            courseStudent.setComments(newComments);
-            courseStudent.setGoogleId(newGoogleId);
-            courseStudent.setTeamName(newTeamName);
-            courseStudent.setSectionName(newSectionName);
+                
+                deleteStudent(courseId, email);
```
Comment (damithc) [inline src/main/java/teammates/storage/api/StudentsDb.java]: Not related to this PR: this can be a cause of errors if the request was terminated before the student was deleted. We should investigate use of datastore level transactions to ensure either both operations happen or both fail.
